moshi/src/main/java/com/squareup/moshi/JsonWriter.kt

```diff
@@ -0,0 +1,553 @@
+   * @return this writer.
+   */
+  @Throws(IOException::class)
+  public abstract fun value(value: Number?): JsonWriter
+
+  /**
+   * Writes `source` directly without encoding its contents. Equivalent to
+   * ```java
+   * try (BufferedSink sink = writer.valueSink()) {
+   *   source.readAll(sink):
+   * }
+   * ```
+   *
+   * @see valueSink
+   */
+  @Throws(IOException::class)
+  public fun value(source: BufferedSource): JsonWriter {
+    check(!promoteValueToName) { "BufferedSource cannot be used as a map key in JSON at path $path" }
+    valueSink().use(source::readAll)
+    return this
+  }
+
+  /**
+   * Returns a [BufferedSink] into which arbitrary data can be written without any additional
+   * encoding. You **must** call [BufferedSink.close] before interacting with this `JsonWriter` instance again.
+   *
+   * Since no validation is performed, options like [serializeNulls] and other writer
+   * configurations are not respected.
+   */
+  @CheckReturnValue
+  @Throws(IOException::class)
+  public abstract fun valueSink(): BufferedSink
+
+  /**
+   * Encodes the value which may be a string, number, boolean, null, map, or list.
+   *
+   * @return this writer.
+   * @see JsonReader.readJsonValue
+   */
+  @Throws(IOException::class)
+  public fun jsonValue(value: Any?): JsonWriter {
+    when (value) {
+      is Map<*, *> -> {
+        beginObject()
+        for ((key, value1) in value) {
+          require(key is String) {
+            if (key == null)
+              "Map keys must be non-null"
+            else
+              "Map keys must be of type String: ${key.javaClass.name}"
+          }
+          name(key)
+          jsonValue(value1)
+        }
+        endObject()
+      }
+      is List<*> -> {
+        beginArray()
+        for (element in value) {
+          jsonValue(element)
+        }
+        endArray()
+      }
+      is String -> value(value as String?)
+      is Boolean -> value(value)
+      is Double -> value(value)
+      is Long -> value(value)
+      is Number -> value(value)
+      null -> nullValue()
+      else -> throw IllegalArgumentException("Unsupported type: ${value.javaClass.name}")
+    }
+    return this
+  }
+
+  /**
+   * Changes the writer to treat the next value as a string name. This is useful for map adapters so
+   * that arbitrary type adapters can use [value] to write a name value.
+   *
+   * In this example, calling this method allows two sequential calls to [value]
+   * to produce the object, `{"a": "b"}`.
+   *
+   * ```java
+   * JsonWriter writer = JsonWriter.of(...);
+   * writer.beginObject();
+   * writer.promoteValueToName();
+   * writer.value("a");
+   * writer.value("b");
+   * writer.endObject();
+   * ```
+   */
+  @Throws(IOException::class)
+  public fun promoteValueToName() {
+    val context = peekScope()
+    if (context != NONEMPTY_OBJECT && context != EMPTY_OBJECT) {
+      throw IllegalStateException("Nesting problem.")
+    }
```

Comment:
Convert this to use `check()`

Reply:
43311aa60ba2b6719e41db95ad9c27885a8f6689